### ISSUE TITLE
ipfs burn-on-read option... facilitates private communication over ipfs

### ISSUE THREAD

**Issue by bobdab** [ipfs/notes#393]
feature request(s):

1) Create a new "ipfs add" command that adds a file to the repository in "burn-on-read" mode.  Perhaps the command would be "ipfs add --burn-on-read".  When the content is read for the first time, it is provided, then deleted (maybe with another option for secure deletion).

This command would facilitate using ipfs for relatively easy, moderately private communications, especially for people who run ipfs on their computers.  I could email you a link with a hash of a message that I added to my own ipfs repo, then after you read it, the message would be deleted.  Users who want to be more secure could add a random string at the end of the message (so that a short message is not easily inferred by an intermediary who has a rainbow table), or the message could be encrypted (GPG/PGP, or otherwise).  

The value of these features compared to using Enigmail (or manual GPG/PGP in an email) is that novice recipients typically do not understand how to keep their private key secure, and they don't understand how to keep their message history secure.  When using regular email with GPG, 5 years from now the recipient's computer could be seized and the key could be used to read all the old email traffic.  With the ipfs burn-on-read option, I could email links to a novice, the novice could open them, the message would appear and then disappear so that the history of communications would be exceedingly difficult to recreate.  All you need is an appropriate ipfs web browser add-in.

Burn-on-read (combined with direct network connections so that the CONTENT of the messages bypasses intermediaries such as email providers) might prove to become extremely important as governments consider mandating backdoors to encryption or encrypted devices (and encrypted email) https://www.theverge.com/2018/9/3/17815196/five-eyes-encryption-backdoors-us-uk-australia-nz-canada.

2) To make this even more secure and faster, two features could be added (if they can not already be acheived in some way)...

ADDED FEATURE 1) I'm not sure about the official policy for announcing to the world that a file has been added to a repo, but the burn-on-read messages should not be announced to the network.  For one thing, a private message is not intended for the whole universe, so Spamming all peers with an announcement is just a waste of bandwidth.  The other thing is that quiet operation improves privacy and reduces the chance that a powerful adversary will detect that the hash exists and try to read the message before the intended recipient reads it.  Burn-on-read would be useless without this feature.

ADDED FEATURE 2, VERSION 1) (Easiest for the recipient of a message to read the message) Establish a different format of an ipfs link that looks specifically at the specified address for the specified hash, maybe something like:  `ipfs://205.123.456.119/tcp/4001/QmNsVsMWYbukZjiBiccMjJABCDEFGHIJKLMNOPQRST ` 
Web browser add-ins would have to capture this and fetch it.

ADDED FEATURE 2, VERSION 2) (Easier to program to provide faster functionality for a variety of applications, but clumsier for novice users) Add a command like this:

`ipfs dht find provs QmXgqKTbzdh83pQtKFb19SpMCpDDcKR2ujqk3pKph9aCNF 205.123.456.789 129.987.545.232 321.989.121.111`

In other words, see if this hash can be found at any of the following addresses (e.g., a server farm that hosts content that the provider wants to be found quickly as soon as the content is created, but the hashes are not announced to the universe).

thx for your consideration here.

**Comment by Stebalien** [comment]
Transferred to notes as this isn't really an actionable issue in go-ipfs (yet).

* The first feature is covered by a "provide nothing" provider strategy (https://github.com/ipfs/go-ipfs/issues/6221).
* The second needs some design work. We need some form of "magnet" link for IPFS that can specify content and one or _more_ sources for that content.
* The third issue isn't really necessary. Given such a magnet link, IPFS could just connect to the referenced node. The providers subsystem doesn't need to be involved.

Note: These issues could all (except "burn on read") be solved with https://github.com/ipfs/research/pull/5.

**Comment by momack2** [comment]
Do you want something like
https://boom.fyi/?

**Comment by bobdab** [comment]
> For start it would be enough to add API for unpin and deleting file From local node.

yes, that would be a start if there is already an easy way to not announce that the file exists (any announcement or visibility would enable snoopers to find the file that was intended to be relatively private).

**Comment by bobdab** [comment]
> Do you want something like https://boom.fyi/?
> […](#)

someone else mentioned shortcut links.  No, that is not interesting to me.  For the system to have real value, it would have to be burn on read.  Here's why....

For a modest degree of privacy, I could email you a link to the real message content.  I could use an email service from a company that does not understand privacy (e.g., Google).  We could send some text message to see if the email provider (or any attacker) is routinely following the links before the reader gets the email.  If testing shows that no spies are reading the links, then I could send you a link, you could open it and read the message, then the server will delete the file automatically.  If someone hacks your email at a later date, all they see is a link to nothing.  Ideally, the file would also have an expiration date to delete the file if it is not read.

There are fancier versions that could be overlaid on a burn-on-read system.

**Comment by bobdab** [comment]
> 1. Provides a false sense of security.

It might be a case of understanding what a burn-on-read system does and does not do.  Let's say I email you a burn-on-read IPFS link using gmail (sender and receiver on gmail).  We run some tests and discover that gmail is not following the links as it harvests information about all users who use its system.  If no interlopers follow the link, then I could send you a link, you could open it, you cold read it, and the message content on IPFS would be deleted instantly.  If your email is hacked at a later date, the spy (or government) sees nothing but invalid links.  If I send you a burn-on-read link, and it is not there when you open it, then somebody is tapping the line (send more test messages to estimate the reliability of the communications channel). 

> 2. Targets a very specific use-case.

yes

> 
> You should consider consider using ephemeral keys for this.

There are plenty of ways to encrypt messages, but I have discovered that hardly anyone is willing to take the time or make the effort to do so... even those who really want to have a conversation without the government (or other people) monitoring every word.  Part of the problem is the need to force the recipient to jump through hoops to read the message.  Maybe if people have IPFS, a simple program could transport the message for the sender and the recipient (with a future generation IPFS browser plugin) could just click on the link and read the message.  Neither the sender nor receiver will have to think too hard about cryptography.

**Comment by RubenKelevra** [comment]
> This is an interesting idea but it:
> 
> 1. Provides a false sense of security.

I agreed, this isn't secure. We cannot avoid that data decoded multiple times.

The only thing we could do which would make sense is to allow just one access and remove the data from the local node and the receiving node afterwards - basically a GC limited to one item.

How useful this is in everyday operations is a different matter.

> In general, we do need private content. However, no design is yet complete enough to be actionable.

ACLs for peers are a great way to implement private content combined with a Web of Trust - which would add specific peers to a trusted list. This can be presented to the user as "friends". This also reduces the bootstrap problem.

Additionally, we could give peers on the "friend" list a specific amount of local space, they can fill with content. This allows a user to save private files on the computers of friends.

I wrote a longer list with advantages of such a system here:
https://github.com/ipfs/notes/issues/397
